jerry-core/api/jerry.c

```diff
@@ -6480,6 +6481,127 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (value);
+
+  if (ecma_get_object_type (obj_p) != ECMA_OBJECT_TYPE_CLASS)
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Value is not a Container or Iterator")));
+  }
+
+  ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) obj_p;
+
+  ecma_collection_t *collection_buffer = ecma_new_collection ();
+  ecma_collection_t *container_p;
+  uint32_t entry_count;
+  uint8_t entry_size;
+
+  uint32_t index = 0;
+  uint8_t iterator_kind = ECMA_ITERATOR__COUNT;
+  ecma_value_t *start_p;
+
+  bool is_key_value_pair = false;
+
+  if (ext_obj_p->u.cls.type == ECMA_OBJECT_CLASS_MAP_ITERATOR
+      || ext_obj_p->u.cls.type == ECMA_OBJECT_CLASS_SET_ITERATOR)
+  {
+    ecma_value_t iterated_value = ext_obj_p->u.cls.u3.iterated_value;
+
+    if (ecma_is_value_empty (iterated_value))
+    {
+      return ecma_op_new_array_object_from_collection (collection_buffer, false);
+    }
+
+    ecma_extended_object_t *map_object_p = (ecma_extended_object_t *) (ecma_get_object_from_value (iterated_value));
+
+    container_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t, map_object_p->u.cls.u3.value);
+    entry_count = ECMA_CONTAINER_ENTRY_COUNT (container_p);
+    index = ext_obj_p->u.cls.u2.iterator_index;
+
+    if (index == entry_count)
+    {
+      return ecma_op_new_array_object_from_collection (collection_buffer, false);
+    }
+
+    entry_size = ecma_op_container_entry_size (map_object_p->u.cls.u2.container_id);
+    start_p = ECMA_CONTAINER_START (container_p);
+
+    iterator_kind = ext_obj_p->u.cls.u1.iterator_kind;
+
+    is_key_value_pair = (iterator_kind == ECMA_ITERATOR_ENTRIES);
+  }
+  else if (jerry_get_container_type (value) != JERRY_CONTAINER_TYPE_INVALID)
+  {
+    container_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t, ext_obj_p->u.cls.u3.value);
+    entry_count = ECMA_CONTAINER_ENTRY_COUNT (container_p);
+    entry_size = ecma_op_container_entry_size (ext_obj_p->u.cls.u2.container_id);
+
+    index = 0;
+    iterator_kind = ECMA_ITERATOR__COUNT;
+    start_p = ECMA_CONTAINER_START (container_p);
+
+    if (ext_obj_p->u.cls.u2.container_id == LIT_MAGIC_STRING_MAP_UL
+        || ext_obj_p->u.cls.u2.container_id == LIT_MAGIC_STRING_WEAKMAP_UL)
+    {
+      is_key_value_pair = true;
+    }
+  }
+  else
+  {
+    ecma_collection_free (collection_buffer);
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Value is not a Container or Iterator")));
+  }
+
+  for (uint32_t i = index; i < entry_count; i+= entry_size)
+  {
+    ecma_value_t *entry_p = start_p + i;
+
+    if (ecma_is_value_empty (*entry_p))
+    {
+      continue;
+    }
+
+    if (iterator_kind != ECMA_ITERATOR_VALUES)
+    {
+      ecma_collection_push_back (collection_buffer, ecma_copy_value_if_not_object (entry_p[0])); //
+    }
+
+    if (is_key_value_pair || iterator_kind == ECMA_ITERATOR_VALUES)
```

Comment:
iterator_kind != ECMA_ITERATOR_KEYS or something?
This enum should have 3 values: keys, values, entries. The latter two only valid for maps, sets only have "keys". 
The first `if` is correct, the second should be simplified.

Reply:
My problem here, and hence the `if` statement, is the following:

- You need to push the `key` on all, except `ECMA_ITERATOR_VALUES`
- You need the `value` on all, except `Set` and `ECMA_ITERATOR_KEY`

And this latter one is the culprit, as it's not a tri-value but a quad, the code - as of now - uses `ECMA_ITERATOR__COUNT` for Sets to distinguish them. Because of this Sets will slip through the `iterator_kind != ECMA_ITERATOR_KEYS` statement and have their "values" pushed as well. The `is_key_value_pair` is there to let only Maps through from the two non-iterator collections. 


Now that i think about it some more, this could be "simplified" down to:
`iterator_kind != ECMA_ITERATOR_KEYS && iterator_kind != ECMA_ITERATOR__COUNT`, which allow the cutting out of the `is_key_value_pair`, if that solution is fine with you.